### PR TITLE
[Artifact Logging PR4] Added log_file_artifact base method

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
Added a base method for `log_file_artifact`, along with a test case.
See [Artifact Logging](https://www.notion.so/Artifact-Logging-API-d3754be3ec874233910cb0c4f23c9d32) for the spec.

### CODE REVIEW

File: composer/core/logging/logger.py [hunk 1]
```diff
@@ -104,29 +115,58 @@ def __init__(
...
             data (LoggerDataDict): The data to log.
         """
-        if isinstance(log_level, str):
-            log_level = LogLevel[log_level.upper()]
+        log_level = LogLevel(log_level)
 
         for destination in self.destinations:
             destination.log_data(self._state, log_level, data)
 
+    def file_artifact(
```
Comment (jbloxham) [inline composer/core/logging/logger.py:131]: I'd like to see this function meet a "VERB_NOUN" naming convention. Maybe just `log_artifact`? `file_artifact` is a bit redundant anyways - aren't all artifacts files?

File: composer/core/logging/logger_destination.py [hunk 2]
```diff
@@ -63,3 +64,40 @@ def log_data(self, state: State, log_level: LogLevel, data: LoggerDataDict):
         """
         del state, log_level, data  # unused
         pass
+
+    def log_file_artifact(
```
Comment (jbloxham) [inline composer/core/logging/logger_destination.py:68]: Would you be OK with shortening this to `log_artifact`?

File: composer/core/logging/logger_destination.py [hunk 3]
```diff
@@ -63,3 +64,40 @@ def log_data(self, state: State, log_level: LogLevel, data: LoggerDataDict):
...
+        self,
+        state: State,
+        log_level: LogLevel,
+        artifact_name: str,
+        file_path: pathlib.Path,
+        *,
+        allow_overwrite: bool,
+    ):
+        """Handle logging of a file artifact stored at ``file_path`` to an artifact named ``artifact_name``.
```
Comment (jbloxham) [inline composer/core/logging/logger_destination.py:77]: Aight I'm starting to get real repetitive now but can we just do a bulk rename of "file artifact" => "artifact"? Or am I dumb and are non-file artifacts a thing

File: composer/core/logging/logger_destination.py [hunk 4]
```diff
@@ -63,3 +64,40 @@ def log_data(self, state: State, log_level: LogLevel, data: LoggerDataDict):
...
+
+        This method is invoked by the :class:`~composer.core.logging.logger.Logger`.
+
+        Implement this method to store the logged file.
```
Comment (jbloxham) [inline composer/core/logging/logger_destination.py:81]: Maybe add a quick note that loggers don't strictly _need_ to implement this method? Some loggers just aren't appropriate for logging whole files.

**Review by jbloxham — APPROVED** [review]
LGTM

File: composer/core/logging/logger.py (hunk 1 above)
Comment (ravi-mosaicml) [reply composer/core/logging/logger.py:131]: Same comment re: https://github.com/mosaicml/composer/pull/698#discussion_r826340994

File: composer/core/logging/logger_destination.py (hunk 2 above)
Comment (ravi-mosaicml) [reply composer/core/logging/logger_destination.py:68]: See below

File: composer/core/logging/logger.py (hunk 1 above)
Comment (ravi-mosaicml) [reply composer/core/logging/logger.py]: Leaving as-is per discussion in https://github.com/mosaicml/composer/pull/698#discussion_r826340994

File: composer/core/logging/logger_destination.py (hunk 3 above)
Comment (ravi-mosaicml) [reply composer/core/logging/logger_destination.py]: Going to leave API as-is for now, so non-file artifacts will not be a future breaking change.

**Pull request merged**